plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/LaunchAppEngineStandardHandlerTest.java

```diff
@@ -0,0 +1,147 @@
+import java.nio.charset.StandardCharsets;
+import org.eclipse.core.commands.ExecutionEvent;
+import org.eclipse.core.commands.ExecutionException;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jst.common.project.facet.core.JavaFacet;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
+import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
+import org.eclipse.wst.common.project.facet.core.ProjectFacetsManager;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.internal.ModuleType;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+
+/**
+ * Tests for the {@link LaunchAppEngineStandardHandler}.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class LaunchAppEngineStandardHandlerTest {
+  private static final IProjectFacetVersion APPENGINE_STANDARD_FACET_VERSION_1 =
+      ProjectFacetsManager.getProjectFacet(AppEngineStandardFacet.ID).getVersion("1");
+
+  @Rule
+  public ServerTracker tracker = new ServerTracker();
+
+  private LaunchAppEngineStandardHandler handler;
+
+  @Rule
+  public TestProjectCreator appEngineStandardProject1 =
+      new TestProjectCreator().withFacetVersions(Lists.newArrayList(JavaFacet.VERSION_1_7,
+          WebFacetUtils.WEB_25, APPENGINE_STANDARD_FACET_VERSION_1));
+  @Rule
+  public TestProjectCreator appEngineStandardProject2 =
+      new TestProjectCreator().withFacetVersions(Lists.newArrayList(JavaFacet.VERSION_1_7,
+          WebFacetUtils.WEB_25, APPENGINE_STANDARD_FACET_VERSION_1));
+
+  private IModule module1;
+  private IModule module2;
+
+  @Before
+  public void setUp() throws CoreException {
+    handler = new LaunchAppEngineStandardHandler();
+    handler.mockLaunch = true;
+  }
+
+  /** Must provide a real project as validators fail. */
+  private static IModule mockAppEngineStandardModule(String serviceId, IProject project)
+      throws CoreException {
+    ModuleType webModuleType = new ModuleType("jst.web", "2.5");
+    IModule module = mock(IModule.class);
+
+    when(module.getName()).thenReturn(serviceId);
+    when(module.getModuleType()).thenReturn(webModuleType);
+    when(module.getProject()).thenReturn(project);
+
+    IFolder webinf = WebProjectUtil.getWebInfDirectory(project);
+    IFile descriptorFile = webinf.getFile("appengine-web.xml");
+    assertTrue(descriptorFile.exists());
+    descriptorFile.setContents(appEngineForService(serviceId), IFile.FORCE, null);
+
+    return module;
+  }
+
+  private static InputStream appEngineForService(String serviceId) {
+    StringBuilder contents = new StringBuilder();
+    contents.append("<?xml version=\"1.0\" encoding=\"utf-8\"?>\n");
+    contents.append("<appengine-web-app xmlns=\"http://appengine.google.com/ns/1.0\">\n");
```

Comment:
You can use single quotes around the attribute values to make the embedded strings more legible.